File: Lv 3/N-Queen.py
answer = 0
def check(x1, y1, x2, y2):
    if abs(y2 - y1) == abs(x2 - x1):
        return True
    else:
        return False

def queen(current, row, n):
    global answer

    if row == n:
        answer += 1
        return
    
    for i in range(n):
        skip = False
        for x1, y1 in current:
            if y1 == i or check(row, i, x1, y1):
                skip = True
                break
                
        if not skip:
            queen(current+[[row, i]], row+1, n)
    
    return

def solution(n):
    global answer
    queen([], 0, n)
    return answer

File: LV 2/N-Queen.java
class Solution {
    int answer = 0;
    int[] col = new int[12];

    public void solve(int n, int row) {
        if (row == n) {
            answer += 1;
        }

        for (int i = 0; i < n; i++) {
            boolean reachable = isReachable(row, i);

            if (!reachable) {
                continue;
            }

            col[row] = i;
            solve(n, row+1);
        }
    }

    private boolean isReachable(int row, int i) {
        boolean reachable = true;
        for (int j = 0; j < row; j++) {
            if (col[j] == i || Math.abs(col[j] - i) == Math.abs(j - row)) {
                reachable = false;
                break;
            }
        }
        return reachable;
    }

    public int solution(int n) {
        solve(n, 0);
        return answer;
    }
}
